### PR TITLE
Add Show Interpolator

### PR DESCRIPTION
**Pull request by ChristopherDavenport**
Having difficulty with the final materialization for sealed traits, as it doesn't find the Show.Shows, which I think it should. 

It's probably something very simple.

Otherwise works as expected.

### CODE REVIEW

**Comment by ritschwumm** [comment]
i don't think you need the implicit conversion to Shows -
why not splice in the calls to .show directly?

**Comment by ritschwumm** [comment]
funny, i had imagined something quite a bit more complicated:
https://github.com/ritschwumm/show-interpolator/blob/master/src/main/scala/ShowMacros.scala?ts=4
you don't need an implicit conversion and an intermediate data type - otoh, it requires macros.

**Comment by ChristopherDavenport** [comment]
Refs #1495

File: core/src/main/scala/scalaz/Show.scala [hunk 1]
```diff
@@ -19,6 +19,19 @@ trait Show[F]  { self =>
 object Show {
   @inline def apply[F](implicit F: Show[F]): Show[F] = F
 
+  trait ContravariantShow[-T] {
+    def shows(t: T): String
+  }
+
+  case class Shows(override val toString: String) extends AnyVal
```
Comment (jdegoes) [inline core/src/main/scala/scalaz/Show.scala]: `final case class ...`

File: core/src/main/scala/scalaz/Show.scala [hunk 2]
```diff
@@ -6,7 +6,7 @@ package scalaz
  * [[scalaz.Cord]] for efficiency.
  */
 ////
-trait Show[F]  { self =>
+trait Show[F] extends Show.ContravariantShow[F]  { self =>
```
Comment (jdegoes) [inline core/src/main/scala/scalaz/Show.scala]: Why is `ContravariantShow` needed?

**Comment by jdegoes** [comment]
Looks like some [compilation errors in the test](https://travis-ci.org/scalaz/scalaz/jobs/290638592).

File: core/src/main/scala/scalaz/Show.scala (hunk 2 above)
Comment (ChristopherDavenport) [reply core/src/main/scala/scalaz/Show.scala]: Without that it infers type `Result1T` not `CopructT` and cannot reach the Shows instance. Should only apply in the string interpolator context. As `Show` is still invariant.

**Comment by ChristopherDavenport** [comment]
I understand that. I'm not seeing the same errors locally with the exact commands in the `.travis.yml`. I'll check again if I'm doing something wrong.

**Comment by ChristopherDavenport** [comment]
`checkGenTypeClasses` regenerates the file, which in this class then lacks the trait extension, it also fails as it doesn't align to what it was previously, as I believe is done [HERE](https://github.com/scalaz/scalaz/blob/89d812642838d90af4c02304ba8ef054f5c95b65/project/build.scala#L160)

So I'm in a battle with the code-generator as the code as written works.

File: core/src/main/scala/scalaz/Show.scala (hunk 2 above)
Comment (TomasMikula) [reply core/src/main/scala/scalaz/Show.scala]: I would say that's an orthogonal problem, not limited to just `Show`. Don't try to solve it in this PR.

**Comment by TomasMikula** [comment]
@ritschwumm I, too, originally thought that it would require macros. After being shown otherwise (by @dwijnand), I now prefer the non-macro implementation, mainly for better readability and lower cost of maintenance.

File: core/src/main/scala/scalaz/Show.scala (hunk 2 above)
Comment (ChristopherDavenport) [reply core/src/main/scala/scalaz/Show.scala]: So leave test pending and remove the `ContravariantShow`?

Comment (TomasMikula) [reply core/src/main/scala/scalaz/Show.scala]: Remove `ContravariantShow`. Either remove the offending test, or add a type annotation: `val r: CoproductT = Result1T`.

**Comment by jdegoes** [comment]
@ChristopherDavenport Looks great, any more changes on your end, or you ready for a merge?

**Comment by ChristopherDavenport** [comment]
@jdegoes  For an initial build it looks good to me. I'll hope we circle around on the contravariant issue shortly, but this looks great in addressing an initial show interpolator.

**Pull request merged**